src/librustc_save_analysis/lib.rs

```diff
@@ -882,6 +909,30 @@ impl<'l, 'tcx: 'l> SaveContext<'l, 'tcx> {
 
         result
     }
+
+    fn make_impl_id(&self, kind: &ImplKind) -> u32 {
+
+        let no = self.impl_counter.get();
+        self.impl_counter.set(no + 1);
+
+        let hasher = DefaultHasher::new();
+
+        hasher.write_u32(no);
+
+        // FIXME. Derive Hash in rls-data.Ô
+        hasher.write_u8(match *kind {
+            ImplKind::Blanket => 0,
+            ImplKind::Deref(_,_) => 1,
+            ImplKind::Direct => 2,
+            ImplKind::Indirect => 3,
+            ImplKind::Inherent => 4,
+        });
+
+        let hash = hasher.finish();
+
+        // is this a good idea?
+        (hash ^ hash << 32) as u32
```

Comment:
You can just use the 64 bit hash - iirc, we use 32 bits for a local id (NodeId) and 64 for a global id (DefId)